src/clj/puppetlabs/services/jruby/jruby_puppet_internal.clj

```diff
@@ -0,0 +1,205 @@
+                                                           "new"
+                                                           (into-array Object
+                                                                       [puppet-config puppet-server-config])
+                                                           JRubyPuppet)
+                        :scripting-container  scripting-container
+                        :environment-registry env-registry})]
+        (.putLast pool instance)
+        instance))))
+
+(schema/defn ^:always-validate
+             get-pool-state :- jruby-schemas/PoolState
+  "Gets the PoolState from the pool context."
+  [context :- jruby-schemas/PoolContext]
+  @(:pool-state context))
+
+(schema/defn ^:always-validate
+  get-pool :- jruby-schemas/pool-queue-type
+  "Gets the JRubyPuppet pool object from the pool context."
+  [context :- jruby-schemas/PoolContext]
+  (:pool (get-pool-state context)))
+
+(schema/defn borrow-without-timeout-fn :- jruby-schemas/JRubyPuppetBorrowResult
+  [pool :- jruby-schemas/pool-queue-type]
+  (.takeFirst pool))
+
+(schema/defn borrow-with-timeout-fn :- jruby-schemas/JRubyPuppetBorrowResult
+  [timeout :- schema/Int
+   pool :- jruby-schemas/pool-queue-type]
+  (.pollFirst pool timeout TimeUnit/MILLISECONDS))
+
+(schema/defn borrow-from-pool!* :- (schema/maybe jruby-schemas/JRubyPuppetInstanceOrRetry)
+  "Given a borrow function and a pool, attempts to borrow a JRuby instance from a pool.
+  If successful, updates the state information and returns the JRuby instance.
+  Returns nil if the borrow function returns nil; throws an exception if
+  the borrow function's return value indicates an error condition."
+  [borrow-fn :- (schema/pred ifn?)
+   ;; it looks unusual that we accept both the pool and the pool-context
+   ;; as arguments, since the PoolContext contains a reference to a pool.  However,
+   ;; in cases such as a full pool flush operation, there may be two distinct
+   ;; pools in play, and the one in the pool-context may be different from the
+   ;; one we're borrowing from.
+   pool :- jruby-schemas/pool-queue-type
+   pool-context :- jruby-schemas/PoolContext]
+  (let [instance (borrow-fn pool)]
+    (cond (instance? PoisonPill instance)
+          (do
+            (.putFirst pool instance)
+            (throw (IllegalStateException.
+                     "Unable to borrow JRuby instance from pool"
+                     (:err instance))))
+
+          (jruby-schemas/jruby-puppet-instance? instance)
+          (do
+            (swap! (:state instance) (fn [m] (update-in m [:request-count] inc)))
+            instance)
+
+          ((some-fn nil? jruby-schemas/retry-poison-pill?) instance)
+          instance
+
+          :else
+          (throw (IllegalStateException.
+                   (str "Borrowed unrecognized object from pool!: " instance))))))
+
+(schema/defn ^:always-validate
+  borrow-from-pool :- jruby-schemas/JRubyPuppetInstanceOrRetry
```

Comment:
Actually, I was thinking `borrow-from-pool!*` should just be `borrow-from-pool*`.  It returns a value, right?